### PR TITLE
[ARTS-330] Gateway server

### PR DESCRIPTION
**Pull request by eladiw**
## Description
1. Adding the SC Gateway server
2. The gateway uses the Discovery server to fetch all registered services and AUTO wire them.
3. Docker-compose updated to support the gateway (however, not mandatory)
4. Added option to switch docker compose from business logic services to all services (bl + config and discovery). using env var. (see md file)

Usage example:
POST http://localhost:8072/practitioner-service/practitioner
[JSON  payload]

### Changes
- [ARTS-330] - Introducing GW
### Checklist:

- [x] Branch name follows convention (feature/arts-#-short-name OR fix/arts-#-short-name)
- [x] I have included a short-meaningful title, description and changes for this PR

[ARTS-330]: https://ndph-arts.atlassian.net/browse/ARTS-330

### CODE REVIEW

File: gateway-server/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+COPY ./${SVC}/src ${SVC}/src
+
+RUN mvn package --projects ${SVC}
+
+FROM openjdk:11.0.9.1-jre-buster
+ARG SVC
+ARG JAR_FILE=${SVC}/target/*.jar
+COPY --from=builder ${JAR_FILE} app.jar
+ENTRYPOINT ["java","-jar","/app.jar"]
+EXPOSE 8761:8761
```
Comment (anatbal) [inline gateway-server/Dockerfile:36]: Why 8761?
```suggestion
EXPOSE 8702:8702
```

File: gateway-server/pom.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-gateway</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-netflix-eureka-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-netflix-eureka-client</artifactId>
```
Comment (anatbal) [inline gateway-server/pom.xml:25]: Why do we need both spring-cloud-starter-netflix-eureka-client and spring-cloud-netflix-eureka-client?

**Review by anatbal — COMMENTED** [review]
Some comments

File: gateway-server/pom.xml (hunk 2 above)
Comment (eladiw) [reply gateway-server/pom.xml:25]: typo

Comment (eladiw) [reply gateway-server/pom.xml]: fixed

File: gateway-server/src/main/resources/bootstrap.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+spring:
+  application:
+    name: gateway-server
+  cloud:
+    gateway:
+      discovery:
+        locator:
+          enabled: true
+          lower-case-service-id: true
```
Comment (tamirkamara) [inline gateway-server/src/main/resources/bootstrap.yml:9]: why? maybe worth to add a comment and explain...

File: gateway-server/src/main/resources/bootstrap.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+server:
+  port: 8072
+
+eureka:
+  client:
+    registerWithEureka: false
+    fetch-registry: true
```
Comment (tamirkamara) [inline gateway-server/src/main/resources/bootstrap.yml:17]: Isn't this the default with the eureka client anyway?

File: README.md [hunk 5]
```diff
@@ -1 +1,14 @@
...
+# A base repo for all microservices
+
+### Running locally using docker compose
+#### Option 1 - All services
+```sh
+export PROFILE=dev
+docker-compose up -d
+```
+
+#### Option 2 - All services in local mode, without service discovery nor config
```
Comment (tamirkamara) [inline README.md:10]: This is the default, so maybe list it first, and then mention you can define a different value to pass to the apps.

File: README.md [hunk 6]
```diff
@@ -1 +1,14 @@
...
+```sh
+export PROFILE=local
```
Comment (tamirkamara) [inline README.md:12]: It looks like this says that setting the env var to local changes how docker-compose brings up services. Is that true? I think that it you call up without names of services everything comes up.

**Review by tamirkamara — CHANGES_REQUESTED** [review]

File: gateway-server/src/main/resources/bootstrap.yml (hunk 4 above)
Comment (eladiw) [reply gateway-server/src/main/resources/bootstrap.yml:17]: yes

File: README.md (hunk 6 above)
Comment (eladiw) [reply README.md:12]: yes not accurate. all are up but not configured...
will rephrase

File: site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/controller/SiteControllerTests.java [hunk 7]
```diff
@@ -29,7 +29,7 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
-@SpringBootTest(properties = { "spring.cloud.config.discovery.enabled = false" , "spring.cloud.config.enabled=false", "server.error.include-message=always", "spring.main.allow-bean-definition-overriding=true" })
+@SpringBootTest(properties = { "spring.cloud.config.discovery.enabled = false" , "spring.cloud.config.enabled=false", "server.error.include-message=always", "spring.main.allow-bean-definition-overriding=true", "fhir.uri=http://localhost:8080" })
```
Comment (tamirkamara) [inline site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/controller/SiteControllerTests.java:32]: looks like saying this active profile is "local" will also remove many of the settings here.

File: docker-compose.yml [hunk 8]
```diff
@@ -63,7 +66,11 @@ services:
       - fhir-api
     environment:
       SERVER_PORT: "8083"
-      SPRING_PROFILES_ACTIVE: "local"
+      SPRING_PROFILES_ACTIVE: ${PROFILE:-local}
+      FHIR_URI: "http://fhir-api:8080"
```
Comment (tamirkamara) [inline docker-compose.yml:70]: setting this env var of fhir_uri takes precedence over the values set in the config file, meaning you couldn't ever change this when running it with compose. is it wise/necessary?
(Applies for the other services in this file)

**Review by tamirkamara — CHANGES_REQUESTED** [review]

File: site-service/src/main/resources/application-local.yml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# for local work we specify the uri's manually using env vars
+server:
+  port: 8083
+  error:
+    include-message: always
```
[the remark this answers is not in the document]
Comment (eladiw) [reply site-service/src/main/resources/application-local.yml:5]: will remove this from local. but I think the application.yml file should not be created here (will be created in the config repo directory)

**Review by tamirkamara — APPROVED** [review]

File: site-service/src/main/resources/application-local.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (eladiw) [reply site-service/src/main/resources/application-local.yml:5]: agreed - to add the yml file in theother repo. leave the state here as is (not adding another yml file) so it wont confuse

File: site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/SiteServiceImplIntegrationTests.java [hunk 10]
```diff
@@ -31,7 +31,7 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @SpringBootTest(properties = { "spring.cloud.config.discovery.enabled = false" , "spring.cloud.config.enabled=false", "server.error.include-message=always", "spring.main.allow-bean-definition-overriding=true" })
```
[the remark this answers is not in the document]
Comment (eladiw) [reply site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/SiteServiceImplIntegrationTests.java]: decided to keep as is. fix in a separate user story

File: site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/controller/SiteControllerTests.java (hunk 7 above)
Comment (eladiw) [reply site-service/src/test/java/uk/ac/ox/ndph/mts/site_service/controller/SiteControllerTests.java:32]: decided to keep as is. fix in a separate user story

File: docker-compose.yml (hunk 8 above)
Comment (eladiw) [reply docker-compose.yml:70]: it is needed to keep both the fhir uri here and in the '-local.yml' file.
if we run the services locally without docker compose, the yml file is needed to set the value.
if running the docker compose then you can set ANY profile and you will get the correct fhir uri. so also needed

**Review by tamirkamara — APPROVED** [review]

**Review by anatbal — APPROVED** [review]

**Pull request merged**